Add documentation for Compose service profiles (#11844)
docker/docker.github.io: compose/profiles.md

---
title: Using profiles with Compose
desription: Using profiles with Compose
keywords: cli, compose, profile, profiles reference
---

> Profiles allow to adjust the Compose application model for various usages and

Reviewer: ```suggestion
Profiles allow adjusting the Compose application model for various usages and
```

> environments by selectively enabling services. For this you can assign a

Reviewer: ```suggestion
environments by selectively enabling services.
```

> service to profiles under which the service should be started only.

Reviewer: ```suggestion
This is achieved by assigning each service to zero or more profiles. If unassigned, the service is _always_ started but if assigned, it is only started if the profile is activated.
```

Reviewer: Not 100 % sure about this change. A little worried that I'm repeating what's said lower down.

> This allows to define additional services in a single `docker-compose.yml` file

Reviewer: ```suggestion
This allows one to define additional services in a single `docker-compose.yml` file
```

that should only be started in specific scenarios, e.g. for debugging or
development tasks.

## Assigning profiles to services

Services are associated with profiles through the
[`profiles` attribute](compose-file/index.md#profiles) which takes an array of
profile names:

```yaml
version: "{{ site.compose_file_v3 }}"
services:
  frontend:
    image: frontend
    profiles: ["frontend"]

  phpmyadmin:
    image: phpmyadmin
    depends_on:
      - db
    profiles:
      - debug

  backend:
    image: backend

  db:
    image: mysql
```

Here the services `frontend` and `phpmyadmin` are assigned to the profiles
`frontend` and `debug` respectively and as such are only started when their
respective profiles are enabled.

> Services without a `profiles` attribute will be _always_ enabled, i.e. in this

Reviewer: ```suggestion
Services without a `profiles` attribute will _always_ be enabled, i.e. in this
```

case running `docker-compose up` would only start `backend` and `db`.

Valid profile names follow the regex format of `[a-zA-Z0-9][a-zA-Z0-9_.-]+`.

> **Note**
>
> The core services of your application should not be assigned `profiles` so
> they will always be enabled and automatically started.

## Enabling profiles

To enable a profile supply the `--profile` [command-line option](reference/overview.md) or
use the [`COMPOSE_PROFILES` environment variable](reference/envvars.md#compose_profiles):

```sh
$ docker-compose --profile debug up
$ COMPOSE_PROFILES=debug docker-compose up
```

> The above would both start your application with the `debug` profile enabled.

Reviewer: ```suggestion
The above command would both start your application with the `debug` profile enabled.
```

> With the above `docker-compose.yml` this would start the services `backend`,

Reviewer: ```suggestion
Using the `docker-compose.yml` file above, this would start the services `backend`,
```

`db` and `phpmyadmin`.

Multiple profiles can be specified by passing multiple `--profile` flags or

> a comma-separated list in `COMPOSE_PROFILES`:

Reviewer: ```suggestion
a comma-separated list for the `COMPOSE_PROFILES` environment variable:
```
